File: 343-integer-break/343-integer-break.py
class Solution:
    def integerBreak(self, n: int) -> int:
        curr = [-inf] * (n + 1)
        curr[0] = 1
        for num in reversed(range(1, n)):
            for remaining in range(num, n + 1):
                curr[remaining] = max(
                    num * curr[remaining - num],
                    curr[remaining]
                )
        return curr[n]

File: 343-integer-break/343-integer-break.java
class Solution {
    public int integerBreak(int n) {
        int[] dp = new int[n + 1];
        dp[0] = 1;
        for(int num = 1; num < n; num++)
            for(int remaining = num; remaining <= n; remaining++)
                dp[remaining] = Math.max(
                    num * dp[remaining - num], 
                    dp[remaining]
                );
        return dp[n];
    }
}